static/js/typing.js

```diff
@@ -0,0 +1,107 @@
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    } else if (operation === 'stop') {
+        if (compose.recipient()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    }
+}
+
+function check_and_send_start() {
+    check_and_send('start');
+}
+
+function get_throttled_function() {
+    return _.throttle(check_and_send_start,
+        TYPING_STARTED_SEND_FREQUENCY,
+        {trailing: false});
+}
+
+function check_and_send_stop() {
+    check_and_send('stop');
+    exports.send_start_notification = get_throttled_function();
+    $(document).on('input', '#new_message_content', exports.send_start_notification);
+}
+
+exports.send_start_notification = get_throttled_function();
+exports.send_stop_notification = _.debounce(check_and_send_stop, TYPING_STOPPED_WAIT_PERIOD);
+
+$(document).on('compose_canceled.zulip', check_and_send_stop);
+$(document).on('compose_finished.zulip', check_and_send_stop);
+$(document).on('blur', '#new_message_content', check_and_send_stop);
+$(document).on('input', '#new_message_content', exports.send_start_notification);
+$(document).on('input', '#new_message_content', exports.send_stop_notification);
+
+function full_name(email) {
+    return people.get_by_email(email).full_name;
+}
+
+exports.hide_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
+        $('#typing_notifications').find('[data-email="' + event.sender.email + '"]').remove();
+        if (users_currently_typing.length === 1) {
+            $('#typing_notifications').hide();
+        }
+        var i = users_currently_typing.indexOf(event.sender.user_id);
+        users_currently_typing.splice(i);
+    }
+};
+
+var debounced_hide_notification = _.debounce(exports.hide_notification,
+  TYPING_STARTED_EXPIRY_PERIOD);
+
+exports.display_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
+        var recipientsMatchingID = event.recipients.filter(function (recipient) {
+            return recipient.user_id === page_params.user_id;
+        });
+        var userIsRecipient = recipientsMatchingID.length > 0;
+        event.sender.name = full_name(event.sender.email);
+        if (narrow.narrowed_to_pms() && userIsRecipient) {
+            $('#typing_notifications').show();
+            if ($('#typing_notifications').find('[data-email="' + event.sender.email + '"]').length === 0) {
+                $('#typing_notifications ul').append(templates.render('typing_notification', event.sender));
+                users_currently_typing.push(event.sender.user_id);
+            }
+            debounced_hide_notification(event);
```

Comment:
Also, can you explain what this logic is supposed to do, e.g. in a comment?  I don't understand under what circumstances the typing notifications will be displayed, and I'm not sure the data structure you're using it right.  We should be keying users by ID, not email, wherever possible, and since the thought was to only display this in PM narrows, it would seem to me that we we'd want the typing notifications to be keyed by both the sender and the recipient (to handle group PMs properly), not just the sender.